walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,65 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.Iterator;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        if (isMeetingRequestTooLong(request)) {
+            return Collections.emptyList();
+        }
+
+        Collection<TimeRange> availableTimeRange = new ArrayList<TimeRange>();
+        availableTimeRange.add(TimeRange.WHOLE_DAY);
+
+        Collection<String> meetingAttendees = request.getAttendees();
+        for (Event e : events) {
+            Collection<String> eventAttendees = e.getAttendees();
+            if (hasCommonAttendees (meetingAttendees, eventAttendees)) {
+                TimeRange eventTime = e.getWhen();
+                updateAvailableTimeRange(availableTimeRange, eventTime, request.getDuration());                
+            }
+        }
+
+        return availableTimeRange;
+    }
+
+    private boolean hasCommonAttendees(Collection<String> meetingAttendees, Collection<String> eventAttendees) {
+        return !Collections.disjoint(meetingAttendees, eventAttendees);
+    }
+
+    private boolean isMeetingRequestTooLong(MeetingRequest request) {
+        return request.getDuration() > TimeRange.WHOLE_DAY.duration();
+    }
+
+    private boolean isTimeslotLongEnough(TimeRange timeslot, long meetingDuration) {
+        return timeslot.duration() >= meetingDuration;
+    }
+
+    private void updateAvailableTimeRange(Collection<TimeRange> availableTimeRange, TimeRange eventTime, long meetingDuration) {
+        for (TimeRange t: new ArrayList<TimeRange>(availableTimeRange)) {
+            if (t.overlaps(eventTime)) {
+                Collection<TimeRange> newTimeslots = splitTimeslot(t, eventTime);
+
+                availableTimeRange.remove(t);
+
+                for (TimeRange newTimeslot : newTimeslots) {
+                    if (isTimeslotLongEnough (newTimeslot, meetingDuration)) {
+                        availableTimeRange.add(newTimeslot);
+                    }
+                }
+            }
+        }
+    }
+
+    private Collection<TimeRange> splitTimeslot(TimeRange availableTimeslot, TimeRange eventTime) {
+        Collection<TimeRange> newTimeslots = new ArrayList<TimeRange>();
+
+        newTimeslots.add(TimeRange.fromStartEnd(availableTimeslot.start(), eventTime.start(), false));
+        newTimeslots.add(TimeRange.fromStartEnd(eventTime.end(), availableTimeslot.end(), false));
+
+        return newTimeslots;
```

Comment:
Alternatively, this can be simplified to
```java
return Arrays.asList(
        TimeRange.fromStartEnd(availableTimeslot.start(), eventTime.start(), false),
        TimeRange.fromStartEnd(eventTime.end(), availableTimeslot.end(), false)
);
```